src/analyzer/analyzerqueue.h

```diff
@@ -1,91 +1,198 @@
+
+    // After enqueuing tracks the analysis must be resumed once.
+    void resume();
+
+    // Cancels a running analysis and discards all enqueued tracks.
+    void cancel();
 
-    void stop();
-    void queueAnalyseTrack(TrackPointer tio);
+  signals:
+    // Progress for individual tracks is passed-through from the workers
+    void trackProgress(TrackId trackId, int analyzerProgress);
+    void progress(int analyzerProgress, int dequeuedCount, int totalCount);
+    void empty(int finishedCount);
+    void done();
 
   public slots:
-    void slotAnalyseTrack(TrackPointer tio);
-    void slotUpdateProgress();
+    void slotAnalyzeTrack(TrackPointer track);
 
-  signals:
-    void trackProgress(int progress);
-    void trackDone(TrackPointer track);
-    void trackFinished(int size);
-    // Signals from AnalyzerQueue Thread:
-    void queueEmpty();
-    void updateProgress();
-
-  protected:
-    void run();
+  private slots:
+    void slotWorkerThreadProgress(int threadId, AnalyzerThreadState threadState, TrackId trackId);
 
   private:
-    struct progress_info {
-        TrackPointer current_track;
-        int track_progress; // in 0.1 %
-        int queue_size;
-        QSemaphore sema;
+    class Worker {
+      public:
+        explicit Worker(std::unique_ptr<AnalyzerThread> thread = std::unique_ptr<AnalyzerThread>())
+            : m_thread(std::move(thread)),
+              m_analyzerProgress(kAnalyzerProgressUnknown),
+              m_threadIdle(false) {
+        }
+        Worker(const Worker&) = delete;
+        Worker(Worker&&) = default;
+
+        operator bool() const {
+            return static_cast<bool>(m_thread);
+        }
+
+        AnalyzerThread* thread() const {
+            DEBUG_ASSERT(m_thread);
+            return m_thread.get();
+        }
+
+        bool threadIdle() const {
+            return m_threadIdle;
+        }
+
+        int analyzerProgress() const {
+            return m_analyzerProgress;
+        }
+
+        void sendNextTrack(TrackPointer track) {
+            DEBUG_ASSERT(track);
+            DEBUG_ASSERT(m_thread);
+            DEBUG_ASSERT(m_threadIdle);
+            m_track = std::move(track);
+            m_threadIdle = false;
+            m_thread->sendNextTrack(m_track);
+        }
+
+        void stopThread() {
+            if (m_thread) {
+                m_thread->stop();
+            }
+        }
+
+        void recvThreadIdle() {
+            DEBUG_ASSERT(m_thread);
+            DEBUG_ASSERT(!m_threadIdle);
+            m_track.reset();
+            m_analyzerProgress = kAnalyzerProgressUnknown;
+            m_threadIdle = true;
+        }
+
+        int recvAnalyzerProgress(TrackId trackId) {
+            DEBUG_ASSERT(m_thread);
+            DEBUG_ASSERT(m_track);
+            DEBUG_ASSERT(m_track->getId() == trackId);
+            DEBUG_ASSERT(!m_threadIdle);
+            return m_analyzerProgress = m_thread->readAnalyzerProgress();
+        }
+
+        void recvThreadExit() {
```

Comment:
I added a comment to AnalyzerThread that we decided not to use queued connections from the main to the worker threads to avoid priority inversion. We only use queued connections (one per thread) in the other direction where it is necessary. As @daschuer mentioned queued connections are not lock-free and AnalyzerThread does not run an event loop.